# Online Retail Customer & Sales Analysis

## Data Understanding

This notebook focuses on understanding the structure, contents, and quality of the Online Retail II dataset before performing any data cleaning or analysis.

## Dataset Description

Source: UCI Machine Learning Repository

Dataset: Online Retail II

Period Covered:
December 2009 – December 2011

The dataset contains transactional records from a UK-based online retail company.

## Objectives of This Notebook

- Load and inspect the dataset
- Understand dataset dimensions
- Review column definitions
- Examine data types
- Identify initial data quality issues
- Document observations before cleaning

## Import Required Libraries

The following libraries are used for data manipulation, numerical operations, and future visualizations.

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)

print("Libraries Loaded Successfully")

Libraries Loaded Successfully


## Load Dataset

The Online Retail II dataset contains two worksheets:

- Year 2009-2010
- Year 2010-2011

Both worksheets will be combined into a single dataset for analysis.

In [2]:
df_2009 = pd.read_excel(
    "../data/raw/online_retail_II.xlsx",
    sheet_name="Year 2009-2010"
)

df_2010 = pd.read_excel(
    "../data/raw/online_retail_II.xlsx",
    sheet_name="Year 2010-2011"
)

df = pd.concat([df_2009, df_2010], ignore_index=True)

## Verify Dataset Loading

Before performing any analysis, we verify that both worksheets have been successfully merged into a single DataFrame.

In [3]:
print(f"Dataset Shape: {df.shape}")

Dataset Shape: (1067371, 8)


## Initial Inspection

The first few rows provide an overview of the dataset structure and help us understand how transactional records are stored.

In [4]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## Random Sample Records

Random observations help reveal patterns that may not appear in the first few rows.

In [5]:
df.sample(5)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
789187,560033,71459,HANGING JAM JAR T-LIGHT HOLDER,12,2011-07-14 13:28:00,0.85,12388.0,Australia
172914,505742,84012,MAGIC SHEEP WOOL GROWING FROM PAPER,12,2010-04-25 16:02:00,2.55,18065.0,United Kingdom
797002,560700,23293,SET OF 12 FAIRY CAKE BAKING CASES,2,2011-07-20 12:38:00,0.83,17229.0,United Kingdom
263320,514830,21755,LOVE BUILDING BLOCK WORD,3,2010-07-06 13:56:00,5.95,12825.0,United Kingdom
61594,494870,21827,EIGHT PIECE CREEPY CRAWLIE SET,4,2010-01-19 12:15:00,1.25,15998.0,United Kingdom


## Dataset Dimensions

The dataset dimensions provide an overview of the number of observations and features available for analysis.

In [6]:
print(f"Number of Rows: {df.shape[0]:,}")
print(f"Number of Columns: {df.shape[1]}")

Number of Rows: 1,067,371
Number of Columns: 8


## Column Overview

The following columns are available in the dataset.

In [7]:
df.columns.tolist()

['Invoice',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'Price',
 'Customer ID',
 'Country']

## Data Types

Understanding data types is essential before performing any cleaning or analysis.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[ns]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 65.1+ MB


## Initial Observations

Based on the dataset information, the following observations were identified:

- The dataset contains 1,067,371 transaction records.
- A total of 8 features are available for analysis.
- The dataset includes numerical, categorical, and datetime variables.
- Missing values are present in some columns and require further investigation.
- The dataset appears to contain customer transaction data from multiple countries.

Customer ID contains a substantial number of missing values and may require special handling during customer-level analysis.

# Data Dictionary

| Column | Description |
|----------|----------|
| Invoice | Unique invoice number for a transaction |
| StockCode | Unique identifier for a product |
| Description | Product description |
| Quantity | Number of units purchased |
| InvoiceDate | Date and time of transaction |
| Price | Unit price of a product |
| Customer ID | Unique customer identifier |
| Country | Country where the customer is located |

In [9]:
missing_values = df.isnull().sum()
missing_percentage = (
    df.isnull().sum() / len(df)
) * 100
pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage": missing_percentage
}).sort_values(
    by="Missing Values",
    ascending=False
)

,Missing Values,Missing Percentage
Customer ID,243007,22.766873
Description,4382,0.410541
StockCode,0,0.000000
Invoice,0,0.000000
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Country,0,0.000000


## Missing Values Assessment

The dataset contains missing values in two columns:

- Customer ID
- Description

Before applying any cleaning techniques, it is important to understand the potential business reasons behind these missing values.

Customer ID contains approximately 22.77% missing values.

These missing values may represent anonymous or unregistered customers rather than data entry errors. Further investigation is required before deciding whether to remove or retain these records.

The Description column contains a small number of missing values (0.41%).

The impact of these missing records is expected to be minimal; however, they should be examined during the data cleaning phase.

In [10]:
df[df["Customer ID"].isnull()].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.00,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.00,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.00,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.00,NaN,United Kingdom
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01 11:49:00,0.55,NaN,United Kingdom
578,489525,85227,SET/6 3D KIT CARDS FOR KIDS,1,2009-12-01 11:49:00,0.85,NaN,United Kingdom
1055,489548,22271,FELTCRAFT DOLL ROSIE,1,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1056,489548,22254,FELT TOADSTOOL LARGE,12,2009-12-01 12:32:00,1.25,NaN,United Kingdom
1057,489548,22273,FELTCRAFT DOLL MOLLY,3,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1058,489548,22195,LARGE HEART MEASURING SPOONS,1,2009-12-01 12:32:00,1.65,NaN,United Kingdom


## Initial Data Quality Findings

Several data quality issues were identified during the initial inspection:

### Missing Customer IDs

Customer IDs are missing in a significant portion of the dataset.

Some transactions appear to be valid purchases despite the absence of customer identifiers, suggesting the possibility of anonymous or guest purchases.

### Negative Quantities

Negative quantities were observed in multiple records.

These records may represent product returns, cancellations, stock corrections, or other business processes rather than data entry errors.

### Zero Prices

Several transactions contain a unit price of zero.

These records require further investigation to determine whether they represent free items, adjustments, promotional transactions, or data quality issues.

### Missing Product Descriptions

A small number of records contain missing product descriptions and should be reviewed during the data cleaning phase.

In [11]:
df[df["Quantity"] < 0].head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


Invoices starting with the letter "C" appear to represent cancelled or returned transactions.

These records contain negative quantities and may reflect product returns or order cancellations rather than standard sales transactions.

In [12]:
cancelled_orders = df[
    df["Invoice"].astype(str).str.startswith("C")
]

print("Cancelled Records:", len(cancelled_orders))

Cancelled Records: 19494


In [13]:
negative_qty = df[
    df["Quantity"] < 0
]

print("Negative Quantity Records:", len(negative_qty))

Negative Quantity Records: 22950


In [14]:
( cancelled_orders.shape[0] == negative_qty.shape[0] )

False

In [15]:
negative_not_cancelled = df[
    (df["Quantity"] < 0)
    &
    (~df["Invoice"].astype(str).str.startswith("C"))
]
negative_not_cancelled.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom


### Investigation Result

Not all negative quantity transactions are associated with invoices starting with "C".

This indicates that negative quantities may represent multiple business processes, including cancellations, returns, stock adjustments, or other operational activities.

Further investigation is required.

### Inventory Adjustment Records

A subset of negative quantity transactions does not correspond to cancelled invoices.

These records frequently exhibit the following characteristics:

- Negative quantities
- Zero unit prices
- Missing customer identifiers
- Operational descriptions such as "lost", "damages", and "short"

This suggests that these records may represent internal inventory adjustments rather than customer transactions.

In [16]:
df[df["Price"] == 0][
    ["Invoice", "StockCode", "Description", "Quantity", "Price", "Customer ID"]
].head(30)

,Invoice,StockCode,Description,Quantity,Price,Customer ID
263,489464,21733,85123a mixed,-96,0.0,NaN
283,489463,71477,short,-240,0.0,NaN
284,489467,85123A,21733 mixed,-192,0.0,NaN
470,489521,21646,NaN,-50,0.0,NaN
3114,489655,20683,NaN,-44,0.0,NaN
3161,489659,21350,NaN,230,0.0,NaN
3162,489660,35956,lost,-1043,0.0,NaN
3168,489663,35605A,damages,-117,0.0,NaN
3731,489781,84292,NaN,17,0.0,NaN
4296,489806,18010,NaN,-770,0.0,NaN


## Price Analysis Findings

Records with a unit price of zero appear to belong to multiple categories:

1. Inventory adjustments and stock corrections
2. Operational records (lost, damaged, short inventory)
3. Potential promotional or free products

Therefore, zero-priced records should not automatically be treated as data quality errors. Further business-context analysis is required before removing them.

In [17]:
df.shape

(1067371, 8)

In [18]:
df.duplicated().sum()

np.int64(34335)

In [19]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,1.067371e+06,1067371,1.067371e+06,824364.000000
mean,9.938898e+00,2011-01-02 21:13:55.394028544,4.649388e+00,15324.638504
min,-8.099500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000
25%,1.000000e+00,2010-07-09 09:46:00,1.250000e+00,13975.000000
50%,3.000000e+00,2010-12-07 15:28:00,2.100000e+00,15255.000000
75%,1.000000e+01,2011-07-22 10:23:00,4.150000e+00,16797.000000
max,8.099500e+04,2011-12-09 12:50:00,3.897000e+04,18287.000000
std,1.727058e+02,NaN,1.235531e+02,1697.464450


# Key Findings

The following observations were identified during the data understanding phase:

- The dataset contains over 1 million transaction records.
- Customer ID has significant missing values (~22.8%).
- Product descriptions contain a small number of missing values.
- Invoices beginning with "C" appear to represent cancellations or returns.
- Negative quantities are not always cancellations and may represent inventory adjustments.
- Zero-priced transactions include both operational adjustments and potential promotional items.
- Extreme values exist in both Quantity and Price fields and require further investigation.
- Additional data cleaning and validation steps are necessary before business analysis.